arch/arm64/src/imx9/imx9_lpspi.c

```diff
@@ -0,0 +1,2060 @@
+
+  DEBUGASSERT(priv != NULL);
+  DEBUGASSERT(priv && priv->spibase);
+  spiinfo("txbuffer=%p rxbuffer=%p nwords=%lu\n", txbuffer, rxbuffer,
+          nwords);
+
+  /* Convert the number of word to a number of bytes */
+
+  nbytes = (priv->nbits > 8) ? nwords << 2 : nwords;
+
+  /* Invalid DMA channels fall back to non-DMA method. */
+
+  if (priv->rxdma == NULL || priv->txdma == NULL
+#ifdef CONFIG_IMX9_LPSPI_DMATHRESHOLD
+      /* If this is a small SPI transfer, then let
+       * imx9_lpspi_exchange_nodma() do the work.
+       */
+
+      || nbytes <= CONFIG_IMX9_LPSPI_DMATHRESHOLD
+#endif
+      )
+    {
+      imx9_lpspi_exchange_nodma(dev, txbuffer, rxbuffer, nwords);
+      return;
+    }
+
+  /* Disable SPI when we are configuring it */
+
+  imx9_lpspi_modifyreg32(priv, IMX9_LPSPI_CR_OFFSET, LPSPI_CR_MEN, 0);
+
+  /* ERR050456 workaround: Reset FIFOs using CR[RST] bit */
+
+  regval = imx9_lpspi_getreg32(priv, IMX9_LPSPI_CFGR1_OFFSET);
+
+  imx9_lpspi_modifyreg32(priv, IMX9_LPSPI_CR_OFFSET,
+                LPSPI_CR_RTF | LPSPI_CR_RRF,
+                LPSPI_CR_RTF | LPSPI_CR_RRF);
+
+  imx9_lpspi_putreg32(priv, IMX9_LPSPI_CFGR1_OFFSET, regval);
+
+  /* Clear all status bits */
+
+  imx9_lpspi_putreg32(priv, IMX9_LPSPI_SR_OFFSET, SPI_SR_CLEAR);
+
+  /* disable DMA */
+
+  imx9_lpspi_putreg32(priv, IMX9_LPSPI_DER_OFFSET, 0);
+
+  if (txbuffer)
+    {
+      up_clean_dcache((uintptr_t)txbuffer, (uintptr_t)txbuffer + nbytes);
+    }
+
+  if (rxbuffer)
+    {
+      up_invalidate_dcache((uintptr_t)rxbuffer,
+                           (uintptr_t)rxbuffer + nbytes);
+    }
+
+  /* Set up the DMA */
+
+  adjust = (priv->nbits > 8) ? 2 : 1;
+
+  struct imx9_edma_xfrconfig_s config;
+
+  config.saddr  = priv->spibase + IMX9_LPSPI_RDR_OFFSET;
+  config.daddr  = (uintptr_t) (rxbuffer ? rxbuffer : rxdummy);
+  config.soff   = 0;
+  config.doff   = rxbuffer ? adjust : 0;
+  config.iter   = nbytes;
+  config.flags  = EDMA_CONFIG_LINKTYPE_LINKNONE;
+  config.ssize  = adjust == 1 ? EDMA_8BIT : EDMA_16BIT;
+  config.dsize  = adjust == 1 ? EDMA_8BIT : EDMA_16BIT;
+  config.nbytes = adjust;
+#ifdef CONFIG_IMX9_EDMA_ELINK
+  config.linkch = NULL;
+#endif
+  imx9_dmach_xfrsetup(priv->rxdma, &config);
+
+  config.saddr  = (uintptr_t) (txbuffer ? txbuffer : &txdummy);
+  config.daddr  = priv->spibase + IMX9_LPSPI_TDR_OFFSET;
+  config.soff   = txbuffer ? adjust : 0;
+  config.doff   = 0;
+  config.iter   = nbytes;
+  config.flags  = EDMA_CONFIG_LINKTYPE_LINKNONE;
+  config.ssize  = adjust == 1 ? EDMA_8BIT : EDMA_16BIT;
+  config.dsize  = adjust == 1 ? EDMA_8BIT : EDMA_16BIT;
+  config.nbytes = adjust;
+#ifdef CONFIG_IMX9_EDMA_ELINK
+  config.linkch = NULL;
+#endif
+  imx9_dmach_xfrsetup(priv->txdma, &config);
+
+  /* Start the DMAs */
+
+  spi_dmarxstart(priv);
+  spi_dmatxstart(priv);
+
+  /* Enable SPI again */
+
+  imx9_lpspi_modifyreg32(priv, IMX9_LPSPI_CR_OFFSET, 0, LPSPI_CR_MEN); /* SPI alkaa lukeen fifoon tässä ? */
```

Comment:
Please remove the comment or translate to English!

Reply:
Hups ;)